### PR TITLE
Add sha256 function

### LINKED ISSUE #1327: Add sha256 function
## Simple Summary
Currently Vyper doesn't 

## Abstract
Add sha256 function, which hcalls 

## Motivation
Currently `sha256` precompile isn't exposed as a function, this VIP adds the functionality.

## Specification

```python
def sha256(bytelike: string) -> bytes32:
      # bytelike can be one of  byte_literal, str_literal, bytes, string, bytes32
      # ...
```

This function will make a call to the pre-compile at  address `0x000...0002`, and return the output as bytes.

## Backwards Compatibility
Feature, forwards compatible.

## Dependencies
No dependencies.

## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/)

### PR DESCRIPTION
**Pull request by jacqueswww**
### What I did
Fixes #1327.

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Comment by jacqueswww** [comment]
@hwwhww So I had managed to get it done today :) would you perhaps be able to test it?

**Comment by hwwhww** [comment]
Thank you so much @jacqueswww! The tests of https://github.com/ethereum/deposit_contract/pull/19 passed with this branch.

**Comment by jacqueswww** [comment]
Cool! @charles-cooper  @fubuloubu I think we are OK merging this? I don't think there is much to discuss on the VIP ?

**Comment by fubuloubu** [comment]
I'm fine with that

File: vyper/functions/functions.py [hunk 1]
```diff
@@ -350,6 +356,93 @@ def _sha3(expr, args, kwargs, context):
...
+            out_len
+        ]
+    ]
+
+
+@signature(('str_literal', 'bytes', 'string', 'bytes32'))
+def sha256(expr, args, kwargs, context):
+    sub = args[0]
+    # Literal input
+    if isinstance(sub, bytes):
```
Comment (charles-cooper) [inline vyper/functions/functions.py]: I think this subroutine has different behavior for string and bytes literals.

File: vyper/functions/functions.py [hunk 2]
```diff
@@ -54,6 +55,11 @@
 )
 
 
+SHA256_ADDRESS = 2
+SHA256_BASE_GAS = 60
+SHA256_PER_WORD_GAS = 6
```
Comment (charles-cooper) [inline vyper/functions/functions.py]: Not sure about this? My reading of the yellow paper says sha256 per word is 12. Seems to be confirmed by https://github.com/ethereum/py-evm/blob/d6adf9039647b6fc39b9d79725e9331008a5c438/eth/constants.py#L85

File: vyper/functions/functions.py (hunk 1 above)
Comment (charles-cooper) [reply vyper/functions/functions.py]: Minimal repro:
```
vyper -f ir /dev/stdin <<EOF
@public
def bar():
    x: bytes32 = sha256(b"test")
    y: bytes32 = sha256("test")
EOF
```

File: vyper/functions/functions.py (hunk 2 above)
Comment (jacqueswww) [reply vyper/functions/functions.py]: Ah right, I must've used the wrong source/reference :+1:

File: vyper/functions/functions.py (hunk 1 above)
Comment (jacqueswww) [reply vyper/functions/functions.py]: Fixed, thanks ;)

**Pull request merged**